File: scripts/mobile_audit.mjs
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { globSync } from "glob";

const interactiveRe = /<(button|a|Link|NavLink|input)\b/;
const targetRe = /(minHeight|minWidth)\s*:\s*(?:'|"|`)?([4-9]\d)/g;
const varRe = /var\(--touch-min\)/;

function auditFile(filePath) {
  const content = readFileSync(filePath, "utf8");
  const lines = content.split(/\r?\n/);
  const issues = [];

  for (let index = 0; index < lines.length; index += 1) {
    if (!interactiveRe.test(lines[index])) continue;

    const window = lines.slice(index, index + 20).join("\n");
    const sizes = [...window.matchAll(targetRe)].map((match) => Number(match[2]));
    const hasTarget = sizes.some((size) => size >= 48) || varRe.test(window);

    if (!hasTarget) {
      issues.push(`${filePath}:${index + 1} interactive element without visible 48px target hint`);
    }
  }

  return issues;
}

const target = process.argv[2] ?? "apps/mobile";
const sourceRoot = globSync(join(target, "src", "**", "*.tsx"), { windowsPathsNoEscape: true });
const issues = sourceRoot.flatMap(auditFile);

if (issues.length > 0) {
  console.log("FAIL");
  for (const issue of issues) {
    console.log(issue);
  }
  process.exit(1);
}

console.log(`PASS: audited ${sourceRoot.length} TSX files under ${target}`);

File: scripts/mobile_audit.py
from __future__ import annotations

import re
import sys
from pathlib import Path

INTERACTIVE_RE = re.compile(r"<(button|a|Link|NavLink|input)\b")
TARGET_RE = re.compile(r"(minHeight|minWidth)\s*:\s*(?:'|\"|`)?([4-9]\d)")
VAR_RE = re.compile(r"var\(--touch-min\)")


def audit_file(path: Path) -> list[str]:
    lines = path.read_text(encoding="utf-8").splitlines()
    issues: list[str] = []

    for idx, line in enumerate(lines):
        if not INTERACTIVE_RE.search(line):
            continue

        window = "\n".join(lines[idx : idx + 20])
        sizes = [int(match.group(2)) for match in TARGET_RE.finditer(window)]
        has_target = any(size >= 48 for size in sizes) or bool(VAR_RE.search(window))

        if not has_target:
            issues.append(f"{path}:{idx + 1} interactive element without visible 48px target hint")

    return issues


def main() -> int:
    target = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("apps/mobile")
    source_root = target / "src" if (target / "src").exists() else target

    if not source_root.exists():
        print(f"ERROR: path not found: {source_root}")
        return 2

    files = sorted(source_root.rglob("*.tsx"))
    issues: list[str] = []

    for file_path in files:
        issues.extend(audit_file(file_path))

    if issues:
        print("FAIL")
        for issue in issues:
          print(issue)
        return 1

    print(f"PASS: audited {len(files)} TSX files under {source_root}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
